text/0000-or-patterns.md

```diff
@@ -0,0 +1,657 @@
+
+## Don't repeat yourself
+
+Consider an example match arm such as (1):
+
+```rust
+Some(Enum::A) | Some(Enum::B) | Some(Enum::C) | Some(Enum::D) => ..
+```
+
+Here, we are repeating `Some($pat)` three times.
+
+Compare (1) to how we could have written this with this RFC (2):
+
+```rust
+Some(Enum::A | Enum::B | Enum::C | Enum::D) => ..
+```
+
+We can see that this is clearly shorter and that the amount of extra work
+we have to do scales linearly with the number of inner variants we mention.
+The ability to nest patterns in this way therefore results in improved
+writing ergonomics.
+
+## Mental model
+
+However, as we know, code is read more than it is written. So are we trading
+readability for increased ergonomics? We believe this is not the case.
+Instead, this RFC aims to improve the readability of code by reducing the
+amount of redundant information that needs to be scanned.
+
+In addition, we aim to more closely align Rust with the mental model that
+*humans* have and how we usually speak and communicate.
+
+Consider that you wanted to ask someone what the *colour* of their *car* was.
+Would you be more inclined to ask:
+
+> Is your car red, white, or blue?
+
+Or would you instead ask:
+
+> Is your car red, your car white, or your car blue?
+
+[CNF]: https://en.wikipedia.org/wiki/Conjunctive_normal_form
+
+[DNF]: https://en.wikipedia.org/wiki/Disjunctive_normal_form
+
+When researching for this RFC; many people were asked and all of them preferred
+the first alternative. This user testing was done on both programmers and
+non-programmers alike and included speakers of: English, German (2) Swedish (3),
+French (2), Portuguese (1), Spanish (2), Farsi (3), Finnish (1), Esperanto (1),
+and Japanese (1).
+
+Thus, we conjecture that it's more common for humans to not distribute and to
+instead use something akin to *conjunctive normal form* ([CNF]) when communicating.
+A likely consequence of this is that a common way to understand snippet (1)
+formulated in *disjunctive normal form* ([DNF]) is to first mentally reconstruct
+it into CNF and then understand the implications of the pattern.
+
+By allowing users to encode their logic in the way they think instead of going
+through more indirect routes, we can improve the understandability of code.
+
+## Reducing complexity with uniformity
+
+A principal way in which programming languages accumulate complexity is by
+adding more and more rules that a programmer needs to keep in their head to
+write or understand what a program does. A consequence of this is that often
+times, caveats and corner cases make for a language that is harder to learn,
+understand, and write in. To avoid such caveats, it is thus imperative that
+we should try to keep the language more uniform rather than less.
+This is an important means through which it becomes possible to give users
+more expressiveness but at the same time limit the cost each feature takes
+from our complexity budget.
```

Comment:
I love this whole paragraph. I'd love to see it added to some top-level documentation of the RFC repository, or some kind of language design philosophy document.